PAS-78: Create Auth Flow

BraidenCutforth opened the pull request:
Auth flow has been added. You should be able to check out the branch, go to `/sign-up` and go through the whole auth process from there. If you find any bugs please comment them! Additionally any other thoughts about behaviour.

BraidenCutforth committed:
PAS-78: Sing-In/Sign-up API hookup

- Add amplify package for signin/signup
- Add exports to auth template, may use in a script
- Add aws-exports.ts file for aws-exports config.
- Add node_modules/ to eslintignore because it was capturing amplify in
  the node modules folder
- Configure amplify in index
- Preliminary hookup to SignInPage

BraidenCutforth committed:
PAS-78 Auth Flow

- Changed to Auth Flow to avoid hosted ui url
- Added necessary explicit auth flows
- Added confirm sign up because of code auth confirmation changed
- Added /confirm-sign-up path for page
- Added better sign up behaviour for errrors
- Added auth hookup to sign in

BraidenCutforth committed:
PAS-78: Fix build error

- Fix Build Error (unused imports)
- Change username to email on sign up page

BraidenCutforth committed:
Add newline to eslintignore

Samantha-norrie commented:
Hi Braiden, sorry if this is out of scope for the ticket, but I noticed a couple of styling things while I went through the sign-in and login:
- 'Already have an account? Sign in here' sometimes is a bit hard to see with certain displays (appears very low), do you think it could benefit from having some padding?
-  We go between using username and email in the sign-up and login pages (I think I may have caused this, sorry!), could just using one help assure that there is no confusion when someone logs in shortly after signing up?

I can't seem to run through the full process of logging in successfully, is there anything you had to do additionally for prep to test? I checked out your branch and did an npm install

Super cool seeing the pages interact together! This was really fun to try out !!

BraidenCutforth commented:
> Hi Braiden, sorry if this is out of scope for the ticket, but I noticed a couple of styling things while I went through the sign-in and login:
> 
> * 'Already have an account? Sign in here' sometimes is a bit hard to see with certain displays (appears very low), do you think it could benefit from having some padding?
> * We go between using username and email in the sign-up and login pages (I think I may have caused this, sorry!), could just using one help assure that there is no confusion when someone logs in shortly after signing up?
> 
> I can't seem to run through the full process of logging in successfully, is there anything you had to do additionally for prep to test? I checked out your branch and did an npm install
> 
> Super cool seeing the pages interact together! This was really fun to try out !!

I am assuming you didn't actually find the sign up page? Currently it is kind of hidden as there isn't a link from the main page, but go to `localhost:3000/sign-up` and try the flow out. It should guide you after that. If there is any confusion, let me know!

isaiahdoyle commented on infrastructure/auth.template:
```diff
@@ -46,7 +46,7 @@
 				],
 				"VerificationMessageTemplate": {
-					"DefaultEmailOption": "CONFIRM_WITH_LINK",
+					"DefaultEmailOption": "CONFIRM_WITH_CODE",
```
Is there any particular reason why confirmation codes are better than confirmation links?

BraidenCutforth replied:
I actually would WAY prefer a confirmation link since you can open it on your phone etc and you don't need to type anything in. That being said, if we allow links right now (raw cognito implementation) we need to use the Cognito hosted ui, which I don't want to do. It makes things a bit more confusing in terms of where people land and some other very specific details. 

But given this, we can actually implement our own link based confirmation process that will actually be even nicer I think. We can use a true templated email (HTML, CSS and the works!) that looks a lot more official.

isaiahdoyle commented on src/aws-exports.ts:
```diff
@@ -0,0 +1,13 @@
+export const config = {
+	Auth: {
+		// identityPoolId: '',
+		// identityPoolRegion: 'us-west-2',
```
Are these lines necessary?

BraidenCutforth replied:
Nope! Good catch, I'll remove!

isaiahdoyle commented on src/pages/accounts/SignInPage.tsx:
```diff
@@ -99,6 +125,7 @@ function SignInPage() {
 				<SignInButton disabled={emailErr} onClick={onSubmit}>
```
Not sure if this is a priority at the moment, but we should definitely look into submitting the form on enter, so someone could potentially log in without the need for a cursor 👀

BraidenCutforth replied:
Its funny because I definitely tried to hit enter to sign in like 4 times. I will make that change now.

isaiahdoyle reviewed (approved):
So cool! Just a couple questions, but it feels so great to be able to interact with the app through more than just the code - not to mention the no-reply email!

Samantha-norrie commented:
> > Hi Braiden, sorry if this is out of scope for the ticket, but I noticed a couple of styling things while I went through the sign-in and login:
> > 
> > * 'Already have an account? Sign in here' sometimes is a bit hard to see with certain displays (appears very low), do you think it could benefit from having some padding?
> > * We go between using username and email in the sign-up and login pages (I think I may have caused this, sorry!), could just using one help assure that there is no confusion when someone logs in shortly after signing up?
> > 
> > I can't seem to run through the full process of logging in successfully, is there anything you had to do additionally for prep to test? I checked out your branch and did an npm install
> > Super cool seeing the pages interact together! This was really fun to try out !!
> 
> I am assuming you didn't actually find the sign up page? Currently it is kind of hidden as there isn't a link from the main page, but go to `localhost:3000/sign-up` and try the flow out. It should guide you after that. If there is any confusion, let me know!

No, I did go to 'localhost:3000/sign-up'. Perhaps I was having some lag, I'll try again sometime tonight/tomorrow! Thanks!

BraidenCutforth committed:
PAS-78: SignIn/SignUp

- Allow pressing enter to submit the form

BraidenCutforth committed:
PAS-78: Use theme colors

- Use theme colors
- Use useReducer hook in SignUp page for some readability advantages,
and allowing multiple property changes one after the other without weird
side effects

BraidenCutforth committed:
PAS-78: Remove unused import

BraidenCutforth committed:
PAS-78: SRP Auth

Samantha-norrie commented on src/pages/accounts/SignUpPage.tsx:
```diff
@@ -47,44 +48,94 @@ const TextLink = styled(TextButton)`
 	if (!password.match(/[A-Z]/)) return false; // Upper-Case
 	if (!password.match(/[a-z]/)) return false; // Lower-Case
 	return true; // Passed all checks
 }
```
The auth flow looks really cool! A small issue that I had happen when I played around with it is that the first password text input box stays in 'error mode' after a faulty password is typed in once and doesn't revert back to being in its regular state even when the input is changed. Don't think this applies to this ticket, but I thought that I should mention it just in case :)

BraidenCutforth replied:
That is definitely part of this ticket! Will fix!

BraidenCutforth committed:
Empty password/email no longer show error
